### ISSUE TITLE
Move away from engine.io (kill it with fire)

### ISSUE THREAD

**Issue by josephg** [rethinkdb/horizon#413]
Context: I'm the author of sharejs (no longer maintained) and [sharedb](https://github.com/share/sharedb) - which solves many of the same problems as horizon (except with OT!). I built the backend for [lever.co](https://www.lever.co/) on top of these systems. And I [had to buid my own socket.io-like library](https://github.com/josephg/node-browserchannel) to work around problems I had with it. So with that in mind:

**Please, say no to socket.io (/engine.io)**.
- Engine.io does not guarantee the order of message delivery. It might seem to - messages _almost always_ arrive in order, but not always. If you move between protocols the ordering of messages can change as different protocols are enabled and disabled.
- This behaviour (at least for the first couple of years of its life) was known but not documented anywhere. I wasted countless hours trying to reproduce these problems in sharejs before I found the culprit.
- Its not needed. [Websockets are supported in every modern browser](http://caniuse.com/#search=websockets).
- If you want to support old browsers [sockjs](https://github.com/sockjs) is a much better option. Its made by much more experienced developers.
- I don't know if this is still true, but socket.io used to have a swathe of problems working through load balancers & clustering.
- You may hear reports about websockets not working through crappy enterprise proxies. These problems all go away if you deploy your site via SSL. (And your site is deployed via ssl, right?). I don't say this glibly - back in 2011 I worked on google wave. The fledgling websocket team in chrome had been running experiments making proto-websocket connections from gmail/chrome to see how well it worked. Websockets over SSL fixed almost all of the proxy issues.

Some of the problems with engine.io may have been fixed in the years since I used it, but based on my technical experience I am extremely nervous about scaling infrastructure on top of engine.io. Websockets are fine. Can we either rip engine.io back out or provide the option to turn it off?

**Comment by flipace** [comment]
I was a bit skeptical when I implemented the engine.io stuff - especially because it seems to be quiet "unmaintained" (e.g. some merges not published to npm...)

Before engine.io pure websockets were used - and it wouldn't be a big issue to revert back to a pure Websocket or SockJS implementation. ( @deontologician i could tackle this again if you'd like )

IF a switch should happen again - I think this should be discussed in tandem with https://github.com/rethinkdb/horizon/issues/358 since this is a critical topic and must probably be handled on a low level.

Big thanks for sharing your experience! Very interesting points!

**Comment by appernetic** [comment]
There is also the problem with the non standard authentication over websockets which is sent in the query string. But it is a general problem not an egine.io specific one...

**Comment by danielmewes** [comment]
Very interesting and useful information @josephg .

> If you move between protocols the ordering of messages can change as different protocols are enabled and disabled.

Do you know if this was/is by design or a bug in the implementation?

Looking at https://github.com/socketio/engine.io-protocol#transport-upgrading :

> the upgrade packet will only be sent once all the buffers of the existing transport are flushed and the transport is considered paused.

This sounds like it would guarantee ordering across transport switches. But maybe it doesn't because of how it's implemented? (or I misunderstand what it means)

**Comment by josephg** [comment]
> This sounds like it would guarantee ordering across transport switches. But maybe it doesn't because of how it's implemented? (or I misunderstand what it means)

Maybe. Maybe that means its fixed? I have no idea! If an XHR message from the client hits the server after the websocket connection is open, does it make it through? What about if a client thinks a message failed (timed out) but it was actually successfully delivered? What about if it gets delivered to another instance of a cluster?

A few years ago running socket.io felt akin to running mongodb nightlies. A new version of Firefox would come out and despite all the protocol fallbacks listed in the documentation your service would just completely stop working until a new version of socket.io came out.

The socket.io team wasn't good enough to fix this problem in the first several years of socket.io's existance. Maybe they've fixed it now? Mongodb seems better now too, but that doesn't mean I'd pick it over rethinkdb or postgres.

If you want IE9 support, Sockjs is actively maintained and run by an extremely smart team (vmware engineers if I recall correctly). Its reliable, and from what I've seen its been rock solid since their first release. Frankly I just don't see any reason to use engine.io when websocket support is so good and sockjs is so stable. Oh yeah, sockjs also doesn't need to make ~3 HTTP roundtrips before connecting via websockets. It starts websockets immediately if they're supported, so in the 95% use case it starts much faster.

**tldr; Horizon should pick one of these choices:**
- Use raw websockets (if you're happy to limit horizon to IE10+)
- SockJS (If IE9 support is important)
- Add an option to pick your streaming protocol. (Devops will love you). Websockets should be the default, but make sock.js available if the user needs IE9. Eventsource would be nice too - its a super clean protocol but it will probably never have IE support. 😢 

I mention devops because in any clustered / load balanced environment you need sticky sessions to make long polling work properly. Thats something you want to think long and hard about - as far as I know AWS ELB still has bugs where clients start bouncing around like mad if the instance they were stuck on goes down.

**Comment by danielmewes** [comment]
One thing that's nice about engine.io over raw websockets (apart from better compatibility with legacy infrastructure) is that it provides heartbeats out of the box. This will be critical for Horizon once we add a user presence API.

sock.js appears to have it too, so that could be an option.

I think we should wait until we get a bit more data on how the current implementation is working out for people. If the engine.io transport ends up actually causing issues for users we should switch.
I personally don't think that it's worth switching right now, since every change like that will likely come with its own set of new considerations which makes it non-trivial. Others on the team might disagree with me about this.
(@deontologician, @dalanmiller, @Tryneus, @coffeemug any opinions?)

**Comment by deontologician** [comment]
I am open to looking into this in the medium term, but for now I think it'll be ok. We just switched to engine.io, so I'd like to not switch back before having some breathing time to think about it.

Websockets are easier from a protocol perspective, more platforms support vanilla websockets out of the box, so it opens us up to Java/Swift clients etc for mobile. I think we'd definitely want to implement heartbeats though, lots of users have run into dead connections and it isn't fun.

Reconnections are unlikely to be solveable by our transport layer alone. We need to be able to restart changefeeds that have been closed on the server and seamlessly catch up what we missed during a disconnection, not just buffer messages. That'll require having knowledge of horizon commands, so we should design it carefully.

**Comment by rationalthug** [comment]
Why not look into using something like [primus](https://github.com/primus/primus)? It supports the WS libraries you've been discussing (and more) and is designed specifically to eliminate issues with switching out the underlying WS implementation. It has support for reconnect, offline detection, etc.

You may also want to have a look at [uWebSockets](https://github.com/alexhultman/uWebSockets) for the server side, which is (apparently) extremely performant and works well with primus/node.js.

**Comment by josephg** [comment]
To be clear on goals:
- What browsers does horizon intend to support?
- What motivated the move away from raw websockets in the first place?

**Comment by josephg** [comment]
> Reconnections are unlikely to be solveable by our transport layer alone. We need to be able to restart changefeeds that have been closed on the server and seamlessly catch up what we missed during a disconnection, not just buffer messages. That'll require having knowledge of horizon commands, so we should design it carefully.

Is there any documentation about horizon's versioning & network model? This might not be the thread to talk about it.. but how I've implemented stuff like this in the past is for a reconnecting client to tell the server "I have subscribed to these queries, and am up to date at version X". Then the server replies with either "Here are changes A,B,C to bring you up to date" or "You are too far out of date. Here's a fresh copy of the query result set".

**Comment by deontologician** [comment]
As far as goals: 
1. We don't have a browser range we want to support at the moment, but pre-websocket browsers will probably be included eventually. Since it's the users of horizon apps that dictate the browser version, and not the developers, we have to make sure we can support users with older browsers.
2. We moved away from websockets because we wanted the fallback options if websockets weren't available, and we wanted heartbeats since people were actually seeing firewalls disconnect them.

As far as your suggestion for  "I have subscribed to these queries, and am up to date at version X", this is more or less what we're planning to do in #9. I'll copy your comment there

**Comment by deontologician** [comment]
Bumping this up. At the very least we should probably move to websockets first

**Comment by danielcompton** [comment]
Just an anecdote on web sockets over SSL, we work with corporate clients and enabling SSL for web sockets helped us get through most of their proxies, but some that do deep packet inspection blocked web sockets (it's unclear if this was intentional).

**Comment by deontologician** [comment]
Yeah, we'll still need some kind of fallback for corporate networks that MitM their employees (gives me the willies, but what are you going to do)

**Comment by macneib** [comment]
another possible contender:
https://github.com/alexhultman/uWebSockets

**Comment by jkarneges** [comment]
FWIW, Meteor uses SockJS and I'm not aware of anyone having major issues with it.

The primary redeeming quality of Engine.IO is that it does transport upgrade rather than transport fallback, which is generally considered preferred behavior.

Personally I kinda hate these "emulate a WebSocket" libraries/protocols because they add extra complexity and make third-party client development confusing (someone using FooLanguage in a non-browser environment has to connect through enginsockiojshackity what now?) but for the greatest browser coverage it's the easiest way forward. The alternative is to create a "clean" HTTP API (you know, based on REST principles or SSE, etc) and write your own transport selection code, but of course that's a heck of a lot more work.

One approach could be to keep the Horizon client using pure WebSockets but allow the user to provide an alternative implementation using dependency injection. This way the code stays clean, but if a user really needs Engine.IO support it would be easy for them to add. Of course there'd still need to be support for this on the server, which could be made optional, and you could also mount the server-side piece on a different path so that it isn't used unless a user means to use it, and also third-party client developers using direct WebSockets wouldn't have to be aware of it.

**Comment by josephg** [comment]
> The primary redeeming quality of Engine.IO is that it does transport upgrade rather than transport fallback, which is generally considered preferred behaviour.

Why?

I mean, doing transport upgrade requires 2-3 extra server roundtrips, and websockets work fine almost everywhere. What do those roundtrips buy you?

**Comment by jkarneges** [comment]
> I mean, doing transport upgrade requires 2-3 extra server roundtrips, and websockets work fine almost everywhere. What do those roundtrips buy you?

Instant connectivity on old browsers at the cost of greater initial overhead on modern browsers. I agree that it's arguable whether this is a good tradeoff anymore in 2016.

**Comment by deontologician** [comment]
Fixed in #526
